### DISCUSSION TITLE
On-demand revalidation works on one domain name but fails on another

### DISCUSSION THREAD

**Discussion by andyzg** [vercel/community#1420] [Help] [answered]
### Summary

When I retrigger a static regeneration of several sites across different custom domains, the sites on the main domain name (the one tied to the production URL) are regenerated successfully but the domain names added to the project are not updated.

I've tried creating a new page that only refreshes when it's revalidated (expires in a week) and then revalidating that page specifically. The logs show that it received a request but the page doesn't update.

Additional context: we're building a website builder and we host user's websites on our domain's subpath but we also support the ability to add a custom domain. Revalidation works for our domain's subpaths but not for their own custom domains.

### Additional information

```
export default async function handler(req, res) {
  // ...some token validation code

  try {
    for (const p of req.body.paths) {
      await res.revalidate(p); // p = `/_site/customdomain/andyzhang.rocks/test-revalidation`
    }

    return res.json({
      success: true,
    });
  } catch (err) {
    return res.status(500).send("Error revalidating");
  }
}
```

The custom domain is mapped to `/_site/customdomain/[host]/test-revalidation.js` which is rewritten through the middleware by using `host` as the custom domain and everything after as the path.

### Example

https://www.andyzhang.rocks/test-revalidation

**Comment by amyegan** [comment, marked as answer]
Thanks for reaching out, Andy. I think the [On-demand Revalidation](https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration#on-demand-revalidation) section of the Next.js On-demand Revalidation doc seems to be related to this situation.

> Note: [Middleware](https://nextjs.org/docs/advanced-features/middleware) won't be executed for On-Demand ISR requests. Instead, call revalidate() on the exact path that you want revalidated. For example, if you have pages/blog/[slug].js and a rewrite from /post-1 -> /blog/post-1, you would need to call res.revalidate('/blog/post-1').

If you revalidate your custom domain's subpath and then visit the custom domain, does the custom domain redirect get the updated content?

**Comment by julesrenaud** [comment]
@amyegan I'm having the exact same issue and it's not related to the middleware not being executed. The revalidated paths are the right ones, but the revalidated content only shows up on the domain that was used to call the /api/revalidate api route (based on the example in the doc). So if I then proceed to go to app.domain.com/revalidated-path (the main domain, the one I used to request the revalidation via app.domain.com/api/revalidate), the up-to-date content is shown. If I go to subdomain.app.domain.com/revalidated-path, the old content is shown.

Both domains are configured in the project and they both exhibit the same behaviour. If I make the /api/revalidate call on the other domain, the same issue is mirrored.

**Reply by amyegan** [reply]
Hi Jules. If you're not using middleware, then it sounds like you have a slightly different problem. Would you mind starting [a new discussion](https://github.com/orgs/vercel/discussions/new?category=help) with the information you shared here plus example links and a public repo/MRE?

**Reply by julesrenaud** [reply]
@amyegan I'm using a middleware, but it's not causing the issue me and the original poster encountered. Revalidation works, but it's only applied to one domain name at a time, the one that was used to call the /api/revalidate route.

Calling subdomain1.domain.com/api/revalidate revalidates content on subdomain1.domain.com but not on subdomain2.domain.com (or any other subdomains that are used on the project). The content stays stale on all the alternate domains for the project.

**Comment by jaisal1024** [comment]
Just for reference, I am encountering the same issues as well. I have 3 domains, I call `revalidate` on only 1 of them and the other 2 have stale content, but the 1 I called revalidate on works. 

If i trigger a re-build, this isn't an issue. All three get valid content.

**Comment by dnishiyama** [comment]
Any update on this? We're running into the same problem.

**Comment by gediminastub** [comment]
Having the same issue. Would love input

**Reply by pv5pm** [reply]
yep same here

**Comment by gediminastub** [comment]
Found how it seems to be solved, trying it out:
https://github.com/vercel/platforms/blob/14a8098ea765560858e81f9629e19780e3ae013c/pages/api/revalidate.ts

Seems that only options is calling api route with revalidation instruction from that domain
